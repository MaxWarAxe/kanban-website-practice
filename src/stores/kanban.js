import { defineStore } from "pinia"
import { ref } from 'vue'
import performersJSON from '../../public/performers.json'
import tasksJSON from '../../public/tasks.json'
import columnsJSON from '../../public/columns.json'

export const useKanbanStore = defineStore('kanban', () => { 
    let columnId = ref(3)
    let taskId = ref(3)
    let performerId = ref(3)
    let columns = ref([])
    let tasks = ref([])
    let performers = ref([])
    let dragging = ref(false)

    function init(){
        // performers.value = performersJSON
        // tasks.value = tasksJSON.map((task)=>{task.performer = performers.value.find(performer => performer.id == task.performerId); return task})
        // columns.value = columnsJSON.forEach((column)=>)
        performers.value = [{id : 1, name : 'Иван Иванчевич'},
                            {id : 2, name : 'Григорий Григориевич'},
                            {id : 3, name : 'Максим Максомов'}]

        tasks.value = [{id : 1, title : 'Задача по проге', content : '1. Сделай то 2. Сделай сё 3. Сделай это', performer: performers.value[0]},
                       {id : 2, title : 'Задача по тестированию', content : 'Напиши тесты по тому, по сему', performer : performers.value[1]},
                       {id : 3, title : 'Вводим в эксплуатацию', content : 'Текст какой-то Текст какой-то Текст какой-то Текст какой-то', performer : performers.value[2]}]
        
        
        columns.value = [{id: 1, header : 'Todo', items : [tasks.value[0]]},
                        {id : 2, header : 'Tests', items : [tasks.value[1]]},
                        {id : 3, header : 'Done', items : [tasks.value[2]]}]
                           console.log(JSON.stringify(columns.value))      
    }

    function removeItemFromColumn(itemID, columnID){
    var column = columns.value.find(column => column.id == columnID)
        column.items = column.items.filter(function(item) {
            return item.id != itemID;
        });
    }

    function addItemToColumn(itemID, columnID){
        var column = columns.value.find(column => column.id == columnID)
        column.items.push(tasks.value.find(task => task.id == itemID))
    }

    function moveItem(itemID, previousColumnID, newColumnID){
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn(){
        columns.value.push({id : ++columnId.value, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID){
        var column = columns.value.find(column => column.id == columnID)
        columns.value = columns.value.filter(function(column) {
            return column.id != columnID;
        });
    }

    function addNewItem(columnID){
        var column = columns.value.find(column => column.id == columnID)
        var newTask = {id : ++taskId.value, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event,itemID,currentColumnID){
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', itemID)
        event.dataTransfer.setData('columnID',currentColumnID)
    }

    function onDropColumn(event, columnID) {
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID, columnID){
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(function(task){
            return task.id != taskID;
        })
    }

    function onRemoveDrop(event){
        const itemID = event.dataTransfer.getData('itemID')
        const previousColumnID = event.dataTransfer.getData('columnID')
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID){
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name){
        performers.value.push({id : ++performerId.value, name : name})
    }

    function removePerformer(performerID){
        tasks.value.forEach(function(task){
            if(task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(function(performer){
            return performer.id != performerID;
        })
    }

    return{
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})