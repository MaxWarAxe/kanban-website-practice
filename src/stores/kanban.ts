import { defineStore } from "pinia"
import { Ref, ref } from 'vue'
import columnsJSON from '@/assets/data/columns.json' 
import type { Column } from '@/classes/Column.ts'
import type { Task } from '@/classes/Task.ts'
import type { Performer } from '@/classes/Performer.ts'

export const useKanbanStore = defineStore('kanban', () => { 
    const columnId : Ref<number> = ref(1)
    const taskId : Ref<number> = ref(1)
    const performerId : Ref<number> = ref(1)
    const columns : Ref<Column[]> = ref(new Array<Column>)
    const tasks : Ref<Task[]> = ref(new Array<Task>)
    const performers : Ref<Performer[]> = ref(new Array<Performer>)
    const dragging : Ref<boolean> = ref(false)

    function init(){   
        columns.value = columnsJSON
        for (let i = 0; i < columns.value.length; i++) {
            for (let j = 0; j < columns.value[i].items.length; j++){
                tasks.value.push(columns.value[i].items[j])
            }
        }
        for (let i = 0; i < tasks.value.length; i++) {
            if (tasks.value[i].performer == null)
                continue
            if (!performers.value.find((performer : Performer)=>performer.id == tasks.value[i].performer.id)) {
                performers.value.push(tasks.value[i].performer)
            }   
        }
        columnId.value = Math.max(...columns.value.map((column : Column) => column.id))
        taskId.value = Math.max(...tasks.value.map((task : Task) => task.id))
        performerId.value = Math.max(...performers.value.map((performer : Performer) => performer.id))
    }

    function removeItemFromColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id === columnID)
        column.items = column.items.filter((item : Task) => item.id !== itemID)
    }

    function addItemToColumn(itemID : number, columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        column.items.push(tasks.value.find((task : Task) => task.id == itemID))
    }

    function moveItem(itemID : number, previousColumnID : number, newColumnID : number) {
        removeItemFromColumn(itemID, previousColumnID)
        addItemToColumn(itemID, newColumnID)
    }

    function addNewColumn() {
        const newColumnID = columnId.value++
        columns.value.push({id : newColumnID, header : 'Новая колонка', items : []})
    }

    function removeColumn(columnID : number) {
        columns.value = columns.value.filter((column : Column) => column.id != columnID)
    }

    function addNewItem(columnID : number) {
        const column = columns.value.find((column : Column) => column.id == columnID)
        const newTaskId = taskId.value++
        const newTask = {id : newTaskId, title : 'Новая задача', content : 'Новый текст', performer : null}
        tasks.value.push(newTask)
        column.items.push(newTask)
    }

    function startDrag(event : DragEvent, itemID : number,currentColumnID : number) {
        dragging.value = true
        event.dataTransfer.dropEffect = 'move'
        event.dataTransfer.effectAllowed = 'move'
        event.dataTransfer.setData('itemID', String(itemID))
        event.dataTransfer.setData('columnID',String(currentColumnID))
    }

    function onDropColumn(event : DragEvent, columnID : number) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        moveItem(itemID,previousColumnID, columnID)
    }    

    function dragEnd(){
        dragging.value = false
    }

    function removeTask(taskID : number, columnID : number) {
        removeItemFromColumn(taskID, columnID)
        tasks.value = tasks.value.filter(task => task.id != taskID)
    }

    function onRemoveDrop(event : DragEvent) {
        const itemID = Number(event.dataTransfer.getData('itemID'))
        const previousColumnID = Number(event.dataTransfer.getData('columnID'))
        removeTask(itemID,previousColumnID)
    }

    function getTask(taskID : number) {
        return tasks.value.find(task => task.id == taskID)
    }

    function addNewPerformer(name : string) {
        const newPerformerId = performerId.value++
        performers.value.push({id : newPerformerId, name : name})
    }

    function removePerformer(performerID : number) {
        tasks.value.forEach(function(task){
            if (task.performer != null && task.performer.id == performerID)
                task.performer = undefined
        })
        performers.value = performers.value.filter(performer=>performer.id != performerID)
    }

    return {
        onRemoveDrop,
        dragging,
        dragEnd,
        columns,
        tasks,
        performers,
        init,
        moveItem,
        addNewColumn,
        removeColumn,
        addNewItem, 
        startDrag,
        onDropColumn,
        getTask,
        addNewPerformer,
        removePerformer
    }
})