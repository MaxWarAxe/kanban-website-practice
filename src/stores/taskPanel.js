import { defineStore } from "pinia"
import { ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    let visible = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {visible,show,close}
})