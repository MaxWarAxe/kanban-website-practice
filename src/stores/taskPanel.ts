import { defineStore } from "pinia"
import { Ref, ref } from 'vue'

export const useTaskPanelStore = defineStore('taskPanel', () => { 
    const visible : Ref<boolean> = ref(false)
    
    function show(){
        visible.value = true
    }

    function close(){
        visible.value = false
    }

    return {
        visible,
        show,
        close
    }
})