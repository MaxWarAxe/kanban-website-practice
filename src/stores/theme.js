import {defineStore} from 'pinia'
import {ref} from 'vue'

export const useThemeStore = defineStore('theme', () => {
    let dark = ref(false)
    let light = ref(false)

    function init(){
        const state1 = localStorageGet()
        if(state1 != undefined){
            dark.value = state1.dark
            light.value = state1.light
            console.log('storageUseed')
        }
        else{
            let darkColorScheme = window.matchMedia('(prefers-color-scheme: dark)')
            dark.value = darkColorScheme.matches
            light.value = !darkColorScheme.matches
            localStorageSave()
        }
        applyClasses()
    }

    function toggleTheme(){
        dark.value = !dark.value
        light.value = !light.value
        applyClasses()
        localStorageSave()
    }

    function applyClasses(){
        document.body.classList.remove('light-theme', 'dark-theme');
        if (dark) {
            document.body.classList.add('dark-theme');
        } else if (light) {
            document.body.classList.add('light-theme');
        }
    }

    function localStorageSave(){
        const stateToSave = {dark : dark.value, 
                            light : light.value}
        localStorage.setItem('THEME_STATE',JSON.stringify(stateToSave));
    }

    function localStorageGet(){
        return JSON.parse(localStorage.getItem('THEME_STATE'))
    }
    
    return {
        dark,
        light,
        init,
        toggleTheme,
        applyClasses,
        localStorageSave,
        localStorageGet
    }
})