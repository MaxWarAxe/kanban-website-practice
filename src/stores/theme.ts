import { defineStore } from 'pinia'
import { Ref, ref } from 'vue'

export const useThemeStore = defineStore('theme', () => {
    const dark : Ref<boolean> = ref(true)
    const light : Ref<boolean> = ref(false)

    function init(){
        const state = localStorageGet()
        if(state != undefined){
            dark.value = state.dark
            light.value = state.light
        }
        else{
            const darkColorScheme = window.matchMedia('(prefers-color-scheme: dark)')
            dark.value = darkColorScheme.matches
            light.value = !darkColorScheme.matches
            localStorageSave()
        }
        applyClasses()
    }

    function toggleTheme(){
        dark.value = !dark.value
        light.value = !light.value
        applyClasses()
        localStorageSave()
    }

    function applyClasses(){
        document.body.classList.remove('light-theme', 'dark-theme');
        if (dark.value) {
            document.body.classList.add('dark-theme');
        }
        if (light.value) {
            document.body.classList.add('light-theme');
        }
    }

    function localStorageSave(){
        const stateToSave = {dark : dark.value, 
                            light : light.value}
        localStorage.setItem('THEME_STATE',JSON.stringify(stateToSave));
    }

    function localStorageGet(){
        return JSON.parse(localStorage.getItem('THEME_STATE'))
    }
    
    return {
        dark,
        light,
        init,
        toggleTheme,
        applyClasses,
        localStorageSave,
        localStorageGet
    }
})