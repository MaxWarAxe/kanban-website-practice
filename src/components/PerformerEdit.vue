<script setup lang="ts">
    import { useKanbanStore } from '@/stores/kanban'
    import { ref, Ref } from 'vue'

    const kanbanStore = useKanbanStore()
    let newPerformerName: Ref<string> = ref('')
</script>

<template>
  <div class="performer-container">
    <div class="performer-container__new-performer-wrapper">
      <input
        class="performer-container__input" 
        type="text" 
        v-model="newPerformerName"
        data-test="name-input"
      >
      <svg
        @click="kanbanStore.addNewPerformer(newPerformerName)"
        class="performer-container__icon" 
        viewBox="0 0 24 24" 
        width="32" 
        height="32"
        data-test="add-button"
      >
        <use href="@/assets/icons/themeIcon.svg#add-user-icon" />
      </svg>
    </div>
      
    <div
      class="performer-container__profile" 
      v-for="performer in kanbanStore.performers"
      :key="performer.id"
    >
      <svg
        @click="kanbanStore.removePerformer(performer.id)" 
        class="performer-container__icon" 
        viewBox="0 0 24 24" 
        width="32" 
        height="32"
      >
        <use href="@/assets/icons/themeIcon.svg#trash-icon" />
      </svg>
      <div class="performer-container__picture">
        <div>{{ performer.name[0] }}</div>
      </div>
      <input
        class="performer-container__performer-name" 
        v-model="performer.name" 
        type="text"
        data-test="performer-change-input"
      >
    </div>
  </div>
</template>

<style lang="scss" scoped>
  .performer-container{
    display: flex;
    padding: 20px;
    gap: 10px;
    flex-direction: column;
    align-items: start;

    &__icon{
      fill: var(--primary-icon-color);
      cursor: pointer;
    }
    &__profile{
      display: flex;
      align-items: center;
      gap: 5px;
    }
    &__picture{
      width: 40px;
      height: 40px;
      display: flex;
      justify-content: center;
      align-items: center;
      font-size: 20px;
      color: var(--button-text-color);
      border-radius: 50%;
      background-color: var(--primary-icon-color);
    }
    &__performer-name{
      font-size: 12px;
      background-color: transparent;
      border: 0px;
      font-family: inter;
      color: var(--primary-text-color);
    }
    &__input{
      background-color: transparent;
      font-size: 18px;
      color: var(--primary-text-color);
    }
    &__new-performer-wrapper{
      display: flex;
      gap: 10px;
    }
  }
</style>