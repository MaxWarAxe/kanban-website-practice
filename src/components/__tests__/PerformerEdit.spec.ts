import { describe,it,expect, vi, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import { setActivePinia,createPinia } from 'pinia'
import { useKanbanStore } from '@/stores/kanban'
import PerformerEdit from '@/components/PerformerEdit.vue'
import { nextTick } from 'vue'

describe('PerformerEdit tests', () => {
    beforeEach(() => {
        // creates a fresh pinia and makes it active
        // so it's automatically picked up by any useStore() call
        // without having to pass it to it: `useStore(pinia)`
        setActivePinia(createPinia())
    })

    it('Коректно рендерится', ()=>{
        const wrapper = mount(PerformerEdit, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn})]
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it('Тест создания исполнителя',async ()=> {
        const wrapper = mount(PerformerEdit, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn , stubActions: false})]
            }
        })
        const kanbanStore = useKanbanStore()
        const addButton = wrapper.get('[data-test="add-button"]')
        addButton.trigger('click')
        await nextTick()
        expect(kanbanStore.performers.length).toBe(1)
    })

    it('Тест создания исполнителя с именем store',async ()=> {
        const wrapper = mount(PerformerEdit, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn , stubActions: false})]
            }
        })
        const kanbanStore = useKanbanStore()
        const input = wrapper.get('[data-test="name-input"]')
        input.setValue('Иван')
        await nextTick()
        const addButton = wrapper.get('[data-test="add-button"]')
        addButton.trigger('click')
        await nextTick()
        expect(kanbanStore.performers[0].name).toBe('Иван')
    })

    it('Тест создания исполнителя с именем (изменение в компоненте)',async ()=> {
        const wrapper = mount(PerformerEdit, {
            global: {
                plugins: [createTestingPinia({createSpy:vi.fn , stubActions: false})]
            }
        })
        const input = wrapper.get('[data-test="name-input"]')
        input.setValue('Иван')
        await nextTick()
        const addButton = wrapper.get('[data-test="add-button"]')
        addButton.trigger('click')
        await nextTick()
        expect(wrapper.get('[data-test="performer-change-input"]').element.value).toBe('Иван')
    })
})