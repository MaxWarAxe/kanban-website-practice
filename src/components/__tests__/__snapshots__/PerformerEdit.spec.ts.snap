// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PerformerEdit tests > Коректно рендерится 1`] = `
"<div data-v-26fb1e78="" class="performer-container">
  <div data-v-26fb1e78="" class="performer-container__new-performer-wrapper"><input data-v-26fb1e78="" class="performer-container__input" type="text" data-test="name-input"><svg data-v-26fb1e78="" class="performer-container__icon" viewBox="0 0 24 24" width="32" height="32" data-test="add-button">
      <use data-v-26fb1e78="" href="/src/assets/icons/themeIcon.svg#add-user-icon"></use>
    </svg></div>
</div>"
`;
